### PR TITLE
Introduce HPA to Rancher API

### PR DESCRIPTION
**Pull request by orangedeng**
Related types PR: https://github.com/rancher/types/pull/775
Related issue: https://github.com/rancher/rancher/issues/19084

### CODE REVIEW

**Comment by jira-sync-svc** [comment]
➤ Kelly Griffin commented:

As noted with PRG‌-650 \- 

This is a feature we need for rancher UI as we can’t see it today.

Numerous customers are/have been asking for this feature.

**Comment by gitlawr** [comment]
@orangedeng If this only works for K8S 1.13+, can you describe the impacts of using v1.12 kubernetes

**Comment by orangedeng** [comment]
Updated. The impact for <v1.12 cluster will be following:
- If users access to project level UI, we will log that we can not find the `horizontalPodAutoscaler.autoscaling.v2beta2` resource.
- You can not find the HPA related schema in the api.

**Comment by gitlawr** [comment]
A 404NotFound error will be logged every ~10 seconds in the subscribe handler of a v1.11 cluster. We need to make an adaptation to previous versions.

**Comment by orangedeng** [comment]
Updated

File: pkg/api/store/hpa/store.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	}
+	return s.Store.Watch(apiContext, schema, opt)
+}
+
+func getClusterName(apiContext *types.APIContext) string {
+	clusterID := apiContext.SubContext["/v3/schemas/cluster"]
+	if clusterID == "" {
+		projectID, ok := apiContext.SubContext["/v3/schemas/project"]
+		if ok {
+			parts := strings.SplitN(projectID, ":", 2)
```
Comment (gitlawr) [inline pkg/api/store/hpa/store.go]: Can we do `clusterID, _ = ref.Parse(projectID)` here

Comment (orangedeng) [reply pkg/api/store/hpa/store.go]: Updated

Comment (orangedeng) [reply pkg/api/store/hpa/store.go]: I am reusing the `ClusterName` function from `clustermanager.Manager`.

**Comment by gitlawr** [comment]
LGTM

**Comment by cjellick** [comment]
LGTM, needs types update/rebase

**Comment by orangedeng** [comment]
Updated

**Comment by cjellick** [comment]
LGTM

**Pull request merged**